go: update to use the new wit-bindgen-go tool and wasip2 target of TinyGo (#170)
bytecodealliance/component-docs: component-model/src/language-support/go.md

# Go Tooling

> The [TinyGo compiler](https://tinygo.org/) v0.33.0 and above has native support for the WebAssembly Component Model and WASI 0.2.0. This guide walks through building a component that implements `example` world defined in the [`add.wit`

Reviewer: While technically TinyGo 0.33.0 has support for `wasip2`, 0.34.0 adds support for `go:wasmexport` and has a number of quality of life fixes for WASI that we might want to say "use 0.34.0" here.

package](../../examples/example-host/add.wit). The component will implement a simple add function.

## 1. Install the TinyGo Compiler and `wasm-tools`

Follow the [TinyGo installation instructions](https://tinygo.org/getting-started/) to install the TinyGo compiler. Additionally, install the `wasm-tools` CLI tool from the [wasm-tools repository](https://github.com/bytecodealliance/wasm-tools/releases).

To verify the installation, run the following commands:

```sh
$ tinygo version
$ wasm-tools -V
```

## 2. Determine which World the Component will Implement

The `wasip2` target of TinyGo requires the imports of `wasi:cli/imports@0.2.0` and thus we need to include them in the `add.wit`. Below is the minimal `add.wit` file that includes the required imports:

```wit
package docs:adder@0.1.0;

> world adder {

Reviewer 2: Can you add a note as to why you needed to add other WASI dependencies to the adder world?

Author: Done

│           ├── adder.exports.go
│           ├── adder.wasm.go
│           ├── adder.wit
│           ├── adder.wit.go
│           └── empty.s
└── wasi
    ├── cli
    │   └── stdout
    │       ├── empty.s
    │       ├── stdout.wasm.go
    │       └── stdout.wit.go
    ├── io
    │   ├── error
    │   │   ├── empty.s
    │   │   ├── error.wit.go
    │   │   └── ioerror.wasm.go
    │   └── streams
    │       ├── empty.s
    │       ├── streams.wasm.go
    │       └── streams.wit.go
    └── random
        └── random
            ├── empty.s
            ├── random.wasm.go
            └── random.wit.go
```

The `adder.exports.go` file contains the exported functions that need to be implemented in the Go code called `Exports`.

## 3. Implement the `add` Function

```Go
package main

import (
	"example.com/internal/example/component/example"
)

func init() {
	example.Exports.Add = func(x int32, y int32) int32 {
		return x + y
	}
}

// main is required for the `wasi` target, even if it isn't used.
func main() {}
```

Go's `init` functions are used to do initialization tasks that
should be done before any other tasks. In this case, we are using it to export the `Add` function and

> make it callable using the generated C bindings (`adder.c`). After populating the `init` function,

Reviewer 2: Looks like c bindings are no longer generated/needed. Should this be updated?

## 4. Build the Component

We can build our component using TinyGo by specifying the wit-package to be `add.wit` and the WIT world to be `adder`. 

> TinyGo will invoke `wasm-tools` to embed the WIT file to the module and componentize it.

Reviewer 2: So this is why wasm-tools is required? This may make it a little clearer 
```suggestion
Under the hood, TinyGo invokes `wasm-tools` to embed the WIT file to the module and componentize it.
```

```sh
$ tinygo build -target=wasip2 -o add.wasm --wit-package add.wit --wit-world adder main.go
```

> We can confirm using the `wasm-tools component wit` command:

Reviewer 2: Can we keep the more verbose wording?

```suggestion
We now have an add component that satisfies our `adder` world, exporting the `add` function, which
we can confirm using the `wasm-tools component wit` command:
```

> ```sh

Reviewer 2: nit: can you use console when displaying both command and output? I know it wasn't doing this properly previously either
```suggestion
```console
```
